---
title: "Analysis of qRT-PCR data from TROST trials 2011-2013"
author: "Heike Sprenger"
date: "Thursday, March 24, 2016"
output:
  html_document:
    highlight: tango
    number_section: yes
    theme: cerulean
    toc: yes
    toc_depth: 4
---

# Set working directory  
```{r set working directory}
getwd()
#setwd("D:/work/repos/trost_qpcr/")
```


# Load workspace, packages and scripts
```{r load workspace, message=FALSE}
# load packages
library(knitr)
library(reshape)
library(pander)
library(pcaMethods)
library(ggplot2)
library(corrplot)
library(gplots)
library(RColorBrewer)

# set options for pander
panderOptions('table.split.table', 200)

# set options for knitr
opts_chunk$set(fig.width=5, fig.height=5, cache=FALSE, highlight = TRUE, fig.show="asis")
opts_knit$set(root.dir = '../')

# load workspace
#load("qpcr_data_trost.RData")
```


# Source R functions
```{r source R functions}
source("../functions/colors.R")
source("../functions/func_prep_pca.R")
source("../functions/func_pca_plots.R")
source("../functions/RemoveFactors_function.R")
source("../functions/func_anova.R")
source("../functions/func_boxplot.R")
```


# Load sample information and modify them
```{r load sample information}
samplelist <- read.table("input/trost/samplelist.txt", header=TRUE, sep="\t")

levels(samplelist$cultivation)

levels(samplelist$trial)

cols_trial_ordered <- cols_trial[c(4:6, 1, 3)]
pie(1:5, col=cols_trial_ordered, cols_trial_ordered)

samplelist$subspecies_id2 <- as.factor(samplelist$subspecies_id)
levels(samplelist$subspecies_id2)[1:4] <- c("Desiree", "Alegria", "Milva", "Saturna")

write.table(samplelist, "output/trost/samplelist.txt", sep="\t")
```


# Load functional annotation
```{r load functional annotation}
assoc_pgsc <- read.table("../trost_transcriptomics/data/PGSC_DM_v3.4_g2t2c2p2func_edit.txt", sep="\t")
colnames(assoc_pgsc) <- c("pgsc_dmg", "pgsc_dmt", "pgsc_dmc", "pgsc_dmp", "func")
head(assoc_pgsc)
```


# Load qpcr raw data (after cleaning) and modify them
```{r load data}
rawdata_cleaned <- read.table("input/trost/rawdata_cleaned.txt", header = T, sep = "\t", row.names = 1)
dim(rawdata_cleaned)
# rows: 92 genes
# columns: 356 samples and 1 additional column with name

# order by rownames = DMT id
rawdata_cleaned <- rawdata_cleaned[order(rownames(rawdata_cleaned)), ]

# merge cleaned rawdata with functional annotation
rawdata_cleaned_merge <- merge(rawdata_cleaned, assoc_pgsc, by.x="row.names", by.y="pgsc_dmt")

# transpose data (only Ct values, without name)
rawdata_cleaned_ct <- t(rawdata_cleaned[,-1])
dim(rawdata_cleaned_ct)
# rows: 356 samples
# columns: 92 genes

# values without HK genes 
HK_index <- which(colnames(rawdata_cleaned_ct) %in% c("PGSC0003DMT400030611", 
                                                      "PGSC0003DMT400068117", 
                                                      "PGSC0003DMT400037585", 
                                                      "PGSC0003DMT400080569"))

rawdata_cleaned_ct_woHK <-rawdata_cleaned_ct[,-HK_index]
dim(rawdata_cleaned_ct_woHK)
# 356 samples, 88 genes

#write.table(colnames(rawdata_cleaned_ct), "output/trost/PGSC_DMT_identifier_qpcr_marker_HK.txt", sep="\t")
```


# Plot ct values
```{r plot ct values}
# boxplot of all ct-values vs. genes, first 4 are HG genes
boxplot.matrix(rawdata_cleaned_ct, use.cols = TRUE, 
               col = c(rep("grey",4), rep("lightblue",88)), 
               las = 2, cex.lab = 0.8)

# plot of HK gene 1 vs. HK gene 2
plot(rawdata_cleaned_ct[,1], rawdata_cleaned_ct[,3])

#class(rawdata_cleaned_ct)
p <- ggplot(as.data.frame(rawdata_cleaned_ct), aes(PGSC0003DMT400030611, PGSC0003DMT400080569))
p + geom_point(alpha =1/3, size=5) 

# histogram of ct-values per gene
pdf("figures/trost/hist_rawdata_cleaned_ct.pdf")
for (i in 1:88){
hist(rawdata_cleaned_ct_woHK[,i], breaks = 30, col = "grey", 
     main = colnames(rawdata_cleaned_ct_woHK)[i])
}
dev.off()

# pairs plot of first 10 genes
pairs(rawdata_cleaned_ct_woHK[,1:10])
```


# Calculate mean of 4 housekeeping genes per sample
```{r calculate mean of 4 housekeeping genes per sample}
housekeeping_mean_all <- apply(rawdata_cleaned_ct[,HK_index], 1, mean, na.rm=TRUE)

# without HK4 !
HK_index_part <- which(colnames(rawdata_cleaned_ct) %in% c("PGSC0003DMT400068117", 
                                                           "PGSC0003DMT400037585", 
                                                           "PGSC0003DMT400080569"))
housekeeping_mean <- apply(rawdata_cleaned_ct[,HK_index_part], 1, mean)
```


# Calculate delta Ct: ``CT_gene - mean(CT_HK)``
```{r calculate delta Ct}
delta_ct <- matrix(1, nrow=nrow(rawdata_cleaned_ct_woHK), ncol=ncol(rawdata_cleaned_ct_woHK))
colnames(delta_ct) <- colnames(rawdata_cleaned_ct_woHK)
rownames(delta_ct) <- rownames(rawdata_cleaned_ct_woHK)

for (i in 1:nrow(rawdata_cleaned_ct_woHK)){
  for (j in 1:ncol(rawdata_cleaned_ct_woHK)){
    delta_ct[i,j] <- rawdata_cleaned_ct_woHK[i,j]-housekeeping_mean_all[i]
    }
  }
```


# Calculate 2 to the power of minus delta Ct
```{r calculate 2 to the power of minus delta Ct}
two_minus_delta_ct <- 2^-(delta_ct)
hist(two_minus_delta_ct, breaks=50, col="grey")

# log 10 transformation of normalized values (2 to the power of minus delta Ct)
log_norm_ct <- log10(two_minus_delta_ct)
hist(log_norm_ct, col="grey", breaks=50)
# low values correpond to low expression in comparison to HK genes (e.g. CT 40)
# high values correpond to high expression in comparison to HK genes (e.g. CT 20)

write.table(log_norm_ct, "output/trost/log_norm_ct.txt", sep="\t")

# log 2 transformation
log2_norm_ct <- log2(two_minus_delta_ct)
log2_norm_ct[1:5,1:5]
two_minus_delta_ct[1:5,1:5]
```


# Compare RNA-Seq results (FPKM values) to qRT-PCR results for marker genes

## Get PGSC information for marker genes (with/without HKG)
```{r get PGSC information for marker genes}
# get DMG identifier
assoc_pgsc_dmg <- droplevels(rawdata_cleaned_merge$pgsc_dmg)
length(assoc_pgsc_dmg)

# without HKG
assoc_pgsc_dmg_woHK <- droplevels(assoc_pgsc_dmg[-HK_index])
length(assoc_pgsc_dmg_woHK)

# get DMT identifier
assoc_pgsc_dmt <- as.factor(rawdata_cleaned_merge$Row.names)

# without HKG
assoc_pgsc_dmt_woHK <- droplevels(assoc_pgsc_dmt[-HK_index])
length(assoc_pgsc_dmt_woHK)

# PGSC table for marker genes (with/without HKG)
assoc_pgsc_marker_HK <- assoc_pgsc[which(assoc_pgsc$pgsc_dmt %in% assoc_pgsc_dmt), ]
assoc_pgsc_marker_woHK <- assoc_pgsc[which(assoc_pgsc$pgsc_dmt %in% assoc_pgsc_dmt_woHK), ]

# order PGSC table by DMG identifier
assoc_pgsc_marker_woHK <- assoc_pgsc_marker_woHK[order(assoc_pgsc_marker_woHK$pgsc_dmg),]
assoc_pgsc_marker_HK <- assoc_pgsc_marker_HK[order(assoc_pgsc_marker_HK$pgsc_dmg),]

dim(assoc_pgsc_marker_HK)
dim(assoc_pgsc_marker_woHK)
```


## Load FPKM values
```{r Load FPKM values}
# load fpkm values (on gene level) --> DMG identifier!
fpkm_genes <- read.table("../trost_transcriptomics/data/genes_FPKMs.txt", header = TRUE, sep = "\t", row.names = 1)

# with housekeeping genes
fpkm_genes_marker_HK <- fpkm_genes[which(rownames(fpkm_genes) %in% assoc_pgsc_dmg), ]

# without housekeeping genes
fpkm_genes_marker_woHK <- fpkm_genes[which(rownames(fpkm_genes) %in% assoc_pgsc_dmg_woHK),]
```


## Combine PGSC annotation with FPKM mean for marker genes
```{r combine PGSC annotation with FPKM mean for marker genes}
# with HKG
fpkm_genes_marker_HK_mean <- apply(fpkm_genes_marker_HK, 1, mean)
fpkm_genes_marker_HK_mean <- fpkm_genes_marker_HK_mean[order(names(fpkm_genes_marker_HK_mean))]
assoc_pgsc_marker_HK <- data.frame(assoc_pgsc_marker_HK, 
                                   fpkm_mean = fpkm_genes_marker_HK_mean)

# without HKG
fpkm_genes_marker_woHK_mean <- apply(fpkm_genes_marker_woHK, 1, mean)
fpkm_genes_marker_woHK_mean <- fpkm_genes_marker_woHK_mean[order(names(fpkm_genes_marker_woHK_mean))]
assoc_pgsc_marker_woHK <- data.frame(assoc_pgsc_marker_woHK, 
                                fpkm_mean = fpkm_genes_marker_woHK_mean)

# order by DMT identifier 
# afterwards, not before, because odering by DMG is different!
assoc_pgsc_marker_HK <- assoc_pgsc_marker_HK[order(assoc_pgsc_marker_HK$pgsc_dmt),]
assoc_pgsc_marker_woHK <- assoc_pgsc_marker_woHK[order(assoc_pgsc_marker_woHK$pgsc_dmt),]
```


## Calculate mean of qRT-PCR results
```{r calculate mean of qRT-PCR results}
# all cultivars
rawdata_cleaned_ct_woHK_mean <- apply(rawdata_cleaned_ct_woHK, 2, mean, na.rm=T)
rawdata_cleaned_ct_HK_mean <- apply(rawdata_cleaned_ct, 2, mean, na.rm=T)
log2_norm_ct_mean <- apply(log2_norm_ct, 2, mean, na.rm=T)

# only check cultivars
log2_norm_ct_check <- subset(log2_norm_ct, samplelist$cultivar %in% c("Alegria", "Desiree", "Milva", "Saturna"))
log2_norm_ct_check_mean <- apply(log2_norm_ct_check, 2, mean, na.rm=T)
```


## Combine PGSC annotation, FPKM mean and Ct mean
```{r combine PGSC annotation, FPKM mean and Ct mean}
assoc_pgsc_marker_woHK = data.frame(assoc_pgsc_marker_woHK, 
                                    ct_mean = rawdata_cleaned_ct_woHK_mean,
                                    log2_norm_ct_mean,
                                    log2_norm_ct_check_mean)

assoc_pgsc_marker_HK = data.frame(assoc_pgsc_marker_HK, 
                                  ct_mean = rawdata_cleaned_ct_HK_mean)


# correlation test
cor.test(assoc_pgsc_marker_HK$ct_mean, log2(assoc_pgsc_marker_HK$fpkm_mean))
cor.test(assoc_pgsc_marker_woHK$log2_norm_ct_mean, log2(assoc_pgsc_marker_woHK$fpkm_mean))
cor.test(assoc_pgsc_marker_woHK$log2_norm_ct_check_mean, log2(assoc_pgsc_marker_woHK$fpkm_mean))
```


## Plot FPKM vs log2 norm Ct values (for check cultivars)
```{r plot FPKM vs log2 norm Ct values (for check cultivars)}
fit <- lm( log2(assoc_pgsc_marker_woHK$fpkm_mean) ~ assoc_pgsc_marker_woHK$log2_norm_ct_check_mean)

plot(assoc_pgsc_marker_woHK$ct_mean, log2(assoc_pgsc_marker_woHK$fpkm_mean), pch=19)
plot(assoc_pgsc_marker_woHK$log2_norm_ct_mean, log2(assoc_pgsc_marker_woHK$fpkm_mean), pch=19)

# pdf("../figures/qpcr_vs_rnaseq.pdf")
# for phd-thesis
pdf("../../Doktorarbeit/figures/qpcr_vs_rnaseq.pdf", width=6, height=6)
par(mar=c(4.5, 4.5, 0.5, 0.5))
plot(assoc_pgsc_marker_woHK$log2_norm_ct_check_mean, 
     log2(assoc_pgsc_marker_woHK$fpkm_mean), 
     pch = 19, cex.lab = 1.5, cex.axis = 1.2, cex = 1.5,
     xlab = bquote("qRT-PCR:" ~ log[2] ~ (2^{-Delta~Ct})), 
     ylab = bquote("RNA-Seq:" ~ log[2] ~ "FPKM"))
abline(fit, lwd = 2, col = "darkgrey")
text(x = -2.5, y = 7.5, "p-value: 1.8e-13", cex = 1.5)
text(x = -2.5, y = 7, "Pearson's correlation: 0.685", cex = 1.5)
dev.off()
```


# PCA
```{r PCA}
log_norm_ct_pareto_rnipals <- func_prep_pca(log_norm_ct, 
                                            scale_method = "pareto", 
                                            center_option = FALSE, 
                                            pc_number = 5, 
                                            pca_method = "rnipals")

log_norm_ct_none_rnipals <- func_prep_pca(log_norm_ct, 
                                          scale_method = "none", 
                                          center_option = FALSE, 
                                          pc_number = 5, 
                                          pca_method = "rnipals")

log_norm_ct_none_ppca <- func_prep_pca(log_norm_ct, 
                                       scale_method = "none", 
                                       center_option = FALSE, 
                                       pc_number = 5, 
                                       pca_method = "ppca")


# export completeObs for later use by RandomForest
write.table(log_norm_ct_none_rnipals@completeObs, "output/trost/log_norm_ct_none_rnipals_completeObs.txt", sep="\t")
```


## PCA plots
```{r PCA plots}
pdf("figures/trost/pca_log_norm_ct.pdf")
# no scaling, rnipals
# treatment
palette(cols_treatment)
func_5pairs_plot(log_norm_ct_none_rnipals, samplelist, 
                 "treatment", 19, "treatment effect (no scaling, rnipals)")
func_pca_plot(log_norm_ct_none_rnipals, 1, 2, samplelist, "treatment", 19, 1, 
              "bottomright", 0.8, maintext = "treatment effect (no scaling, rnipals)")

# cultivation
palette(cols_sample_time2)
func_5pairs_plot(log_norm_ct_none_rnipals, samplelist, 
                 "cultivation", 19, "cultivation effect (no scaling, rnipals)")
func_pca_plot(log_norm_ct_none_rnipals, 1, 2, samplelist, "cultivation", 19, 1, 
              "bottomright", 0.8, maintext = "cultivation effect (no scaling, rnipals)")

# trial
palette(cols_trial_ordered)
func_5pairs_plot(log_norm_ct_none_rnipals, samplelist, 
                 "trial", 19, "trial effect (no scaling, rnipals)")

plot(log_norm_ct_none_rnipals@scores[,1], 
     log_norm_ct_none_rnipals@scores[,2], 
     col = samplelist$trial, pch=19)
legend("topleft", fill = 1:5, 
       c("JKI field 2012", "MPI field 2011", "MPI field 2012", "MPI-MP GH3", "MPI-MP GH5"))

plot(log_norm_ct_none_rnipals@scores[,3], 
     log_norm_ct_none_rnipals@scores[,4], 
     col = samplelist$trial, pch = 19, 
     xlim = c(-4.5, 4.5), xlab = "PC3", ylab = "PC4", cex.lab = 1.5, cex.axis = 1.2)
legend("bottomright", fill = 1:5, 
       c("JKI field 2012", "MPI field 2011", "MPI field 2012", "MPI-MP GH3", "MPI-MP GH5"))

# cultivar
palette(rainbow(34))
func_5pairs_plot(log_norm_ct_none_rnipals, samplelist, 
                 "cultivar", 19, "cultivar effect (no scaling, rnipals)")

plot(log_norm_ct_none_rnipals@scores[,1], 
     log_norm_ct_none_rnipals@scores[,3], 
     col = samplelist$cultivar, pch = 19)

text(log_norm_ct_none_rnipals@scores[,1], 
     log_norm_ct_none_rnipals@scores[,3], 
     labels = samplelist$cultivar, cex = 0.7)

plot(log_norm_ct_none_rnipals@scores[,2], 
     log_norm_ct_none_rnipals@scores[,3], 
     col = samplelist$cultivar, pch = 19)
text(log_norm_ct_none_rnipals@scores[,2], 
     log_norm_ct_none_rnipals@scores[,3], 
     labels = samplelist$cultivar, cex = 0.7)


# pareto scaling, rnipals
# treatment
palette(cols_treatment)
func_5pairs_plot(log_norm_ct_pareto_rnipals, samplelist, 
                 "treatment", 19, "treatment effect (pareto scaling, rnipals)")
func_pca_plot(log_norm_ct_pareto_rnipals, 1, 2, samplelist, "treatment", 19, 1, 
              "bottomright", 0.8, maintext = "treatment effect (pareto scaling, rnipals)")

# trial
palette(cols_trial_ordered)
func_5pairs_plot(log_norm_ct_pareto_rnipals, samplelist, 
                 "trial", 19, "trial effect (pareto scaling, rnipals)")
func_pca_plot(log_norm_ct_pareto_rnipals, 1, 2, samplelist, "trial", 19, 1, 
              "bottomright", 0.8, maintext = "trial effect (pareto scaling, rnipals)")

# cultivar
palette(rainbow(34))
func_5pairs_plot(log_norm_ct_pareto_rnipals, samplelist,
                 "cultivar", 19, "cultivar effect (pareto scaling, rnipals)")
func_pca_plot(log_norm_ct_pareto_rnipals, 1, 4, samplelist, "cultivar", 19, 1, 
              "bottomright", 0.8, maintext = "cultivar effect (pareto scaling, rnipals)")

plot(log_norm_ct_pareto_rnipals@scores[,1], 
     log_norm_ct_pareto_rnipals@scores[,4], 
     col = samplelist$cultivar, pch = 19, 
     xlab = "PC1 (25%)", ylab = "PC4 (5%)")

text(log_norm_ct_pareto_rnipals@scores[,1], 
     log_norm_ct_pareto_rnipals@scores[,4], 
     labels = samplelist$cultivar, cex = 0.7)


# no scaling, ppca
# treatment
palette(cols_treatment)
func_5pairs_plot(log_norm_ct_none_ppca, samplelist, 
                 "treatment", 19, "treatment effect (no scaling, ppca)")

# trial
palette(cols_trial_ordered)
func_5pairs_plot(log_norm_ct_none_ppca, samplelist, 
                 "trial", 19, "trial effect (no scaling, ppca)")

func_pca_plot(log_norm_ct_none_ppca, 1, 2, 
              samplelist, "trial", 19, 1, 
              "bottomright", 0.8, 
              maintext = "trial effect (no scaling, ppca)")

# cultivar
palette(rainbow(34))
func_5pairs_plot(log_norm_ct_none_ppca, samplelist, 
                 "cultivar", 19, "cultivar effect (no scaling, ppca)")
# separation of some cultivars in PC2 vs. PC3
func_pca_plot(log_norm_ct_none_ppca, 2, 3, 
              samplelist, "cultivar", 19, 1, 
              "bottomright", 0.8, 
              maintext = "cultivar effect (no scaling, ppca)")

plot(log_norm_ct_none_ppca@scores[,2], 
     log_norm_ct_none_ppca@scores[,3], 
     col = samplelist$cultivar, pch = 19)

text(log_norm_ct_none_ppca@scores[,2], 
     log_norm_ct_none_ppca@scores[,3], 
     labels = samplelist$cultivar, cex = 0.7)
dev.off()

palette("default")
```


## PCA all samples for thesis
```{r PCA all samples for thesis}
log_norm_ct_none_rnipals@R2

# trials
#pdf("../../Doktorarbeit/figures/pca_log_norm_ct_trial.pdf", width = 6, height = 6)
par(mar=c(4.5, 4.5, 0.5, 0.5))
palette(cols_trial_ordered)
plot(log_norm_ct_none_rnipals@scores[,3],  
     log_norm_ct_none_rnipals@scores[,4], 
     col = samplelist$trial, pch = 19, 
     xlim = c(-4.5, 4.5), xlab = "PC3 (8.6%)", ylab = "PC4 (6.9%)", 
     cex.lab = 1.5, cex.axis = 1.2, cex = 1.5)
legend("bottomright", fill = 1:5, cex = 1.2, bty = "n",
       c("JKI field 2012", "MPI field 2011", "MPI field 2012", "MPI-MP GH3", "MPI-MP GH5"))
#dev.off()

# treatment
#pdf("../../Doktorarbeit/figures/pca_log_norm_ct_treatment.pdf", width = 6, height = 6)
par(mar=c(4.5, 4.5, 0.5, 0.5))
palette(cols_treatment)
plot(log_norm_ct_none_rnipals@scores[,1], 
     log_norm_ct_none_rnipals@scores[,4], 
     col = samplelist$treatment, pch = 19, 
     xlab = "PC1 (23.4%)", ylab = "PC4 (6.9%)", 
     cex.lab = 1.5, cex.axis = 1.2, cex = 1.5)
legend("bottomleft", fill = 1:2, c("control", "drought stress"), cex = 1.2, bty = "n")
#dev.off()

# cultivars
#pdf("../../Doktorarbeit/figures/pca_log_norm_ct_cultivar.pdf", width = 6, height = 6)
par(mar=c(4.5, 4.5, 0.5, 0.5))
palette(rainbow(34))
plot(log_norm_ct_none_rnipals@scores[,2], 
     log_norm_ct_none_rnipals@scores[,3], 
     col = samplelist$cultivar, pch=19, 
     xlim = c(-4.5, 4.5), xlab = "PC2 (14.2%)", ylab = "PC3 (8.6%)", 
     cex.lab = 1.5, cex.axis = 1.2, cex = 1.5, ylim = c(-4.5, 4.5))
legend("topleft", fill = 1:34, levels(samplelist$subspecies_id2), 
       cex = 0.8, horiz = F, ncol = 7, x.intersp = 0.2)
#dev.off()
```


# Only check cultivars
```{r only check cultivars}
samplelist_check <- subset(samplelist, samplelist$cultivar %in% c("Alegria", "Desiree", "Milva", "Saturna"))
samplelist_check$cultivar <- factor(samplelist_check$cultivar, levels=c("Alegria", "Milva", "Desiree", "Saturna"))

log_norm_ct_check <- subset(log_norm_ct, samplelist$cultivar %in% c("Alegria", "Desiree", "Milva", "Saturna"))

log_norm_ct_check_none_rnipals <- func_prep_pca(log_norm_ct_check, 
                                                scale_method = "none", 
                                                center_option = FALSE, 
                                                pc_number = 5, 
                                                pca_method = "rnipals")

# no scaling, rnipals
pdf("figures/trost/pca_log_norm_ct_check_cultivars.pdf")
par(mar=c(4.5, 4.5, 0.5, 0.5))
palette(cols_cultivar2)
pairs(log_norm_ct_check_none_rnipals@scores[,1:5], 
      col = samplelist_check$treatment, pch = 19)
pairs(log_norm_ct_check_none_rnipals@scores[,1:5], 
      col = samplelist_check$trial, pch = 19)
pairs(log_norm_ct_check_none_rnipals@scores[,1:5], 
      col = samplelist_check$cultivar, pch = 19)

plot(log_norm_ct_check_none_rnipals@scores[,1], 
     log_norm_ct_check_none_rnipals@scores[,2], 
     col = samplelist_check$cultivar, pch = 19)
legend("topright", levels(samplelist_check$cultivar), fill = 1:4)

plot(log_norm_ct_check_none_rnipals@scores[,1], 
     log_norm_ct_check_none_rnipals@scores[,3], 
     col = samplelist_check$cultivar, pch=19, 
     xlab = "PC1 (44%)", ylab = "PC3 (13%)", 
     cex = 2, cex.lab = 1.5, cex.axis = 1.2)
legend("bottomright", levels(samplelist_check$cultivar), fill = 1:4, cex = 1.5)
dev.off()

# for phd-thesis
#pdf("../../Doktorarbeit/figures/pca_log_norm_ct_check_cultivars.pdf", width=6, height=6)
par(mar=c(4.5, 4.5, 0.5, 0.5))
palette(cols_cultivar2)
plot(log_norm_ct_check_none_rnipals@scores[,1], 
     log_norm_ct_check_none_rnipals@scores[,3], 
     col = samplelist_check$cultivar, pch = 19, 
     xlab = "PC1 (44.7%)", ylab = "PC3 (13.0%)", 
     cex= 2, cex.lab = 1.5, cex.axis = 1.2)
legend("bottomright", levels(samplelist_check$cultivar), 
       fill = 1:4, cex = 1.2, bty = "n")
#dev.off()

head(sort(log_norm_ct_check_none_rnipals@loadings[,1]))
tail(sort(log_norm_ct_check_none_rnipals@loadings[,1]))

biplot(log_norm_ct_check_none_rnipals)
```


# Correlation plot
```{r corrplot}
cor_log_norm_ct <- cor(log_norm_ct,use="complete")
write.table(cor_log_norm_ct, "output/trost/cor_log_norm_ct.txt", sep="\t")

pdf("figures/trost/cor_log_norm_ct.pdf", width=10, height=10)
corrplot(cor_log_norm_ct, method="color", order="hclust", hclust.method="average", tl.col="black", tl.cex=0.5)
dev.off()
```


# Remove Factors by ANOVA models using R script by Jan Lisec
```{r normalize (Remove Factors)}
levels(samplelist$trial)
levels(samplelist$treatment)
levels(samplelist$cultivar)

# treatment included
log_norm_ct_2 <- apply(log_norm_ct, 2, RemoveFactors, sam=samplelist, 
                              facs=c("cultivar", "trial"), 
                              keep=c("cultivar"))

# treatment NOT included!
log_norm_ct_3 <- apply(log_norm_ct, 2, RemoveFactors, sam=samplelist, 
                              facs=c("cultivar", "trial", "treatment"), 
                              keep=c("cultivar", "treatment"))
```


## PCA after Remove Factors
```{r PCA after Remove Factors}

log_norm_ct_pareto_rnipals_2 <- func_prep_pca(log_norm_ct_2, 
                                              scale_method = "pareto", 
                                              center_option = FALSE, 
                                              pc_number = 5, 
                                              pca_method = "rnipals")

log_norm_ct_pareto_rnipals_3 <- func_prep_pca(log_norm_ct_3, 
                                              scale_method = "pareto", 
                                              center_option = FALSE, 
                                              pc_number = 5, 
                                              pca_method = "rnipals")

pdf("figures/trost/pca_log_norm_ct_2.pdf")

# pareto scaling, rnipals
# treatment
palette("default")
pairs(log_norm_ct_pareto_rnipals_2@scores[,1:5], 
      col = samplelist$treatment, pch=19) # --> separation in PC1 vs.PC3

plot(log_norm_ct_pareto_rnipals_2@scores[,1], 
     log_norm_ct_pareto_rnipals_2@scores[,3], 
     col = samplelist$treatment, pch = 19, 
     xlab = "PC1 (25%)", ylab = "PC3 (8%)")
legend("bottomright", levels(samplelist$treatment), fill=1:2)

# trial
pairs(log_norm_ct_pareto_rnipals_2@scores[,1:5], 
      col = samplelist$trial, pch=19)

plot(log_norm_ct_pareto_rnipals_2@scores[,1], 
     log_norm_ct_pareto_rnipals_2@scores[,2], 
     col = samplelist$trial, pch = 19, 
     xlab = "PC1 (25%)", ylab = "PC2 (8%)")

legend("topleft", levels(samplelist$trial), fill=1:5)

# cultivars
palette(rainbow(34))
pairs(log_norm_ct_pareto_rnipals_2@scores[,1:5], 
      col = samplelist$cultivar, pch = 19)

plot(log_norm_ct_pareto_rnipals_2@scores[,3], 
     log_norm_ct_pareto_rnipals_2@scores[,4], 
     col=samplelist$cultivar, pch = 19, 
     xlab = "PC1 (25%)", ylab = "PC4 (5%)")

text(log_norm_ct_pareto_rnipals_2@scores[,3], 
     log_norm_ct_pareto_rnipals_2@scores[,4], 
     labels = samplelist$cultivar, cex=1)

dev.off()
palette("default")
```



# Only field trials (same in qpcr and rnaseq)
## t01, t05, t06
```{r only field trials}
# FPKM values of field trials (without housekeeping genes)
fpkm_genes_marker_field <- fpkm_genes_marker_woHK[, c(1:8,33:48)]
dim(fpkm_genes_marker_field)

fpkm_genes_marker_field_mean <- apply(fpkm_genes_marker_field, 1, mean)
fpkm_genes_marker_field_mean <- fpkm_genes_marker_field_mean[order(names(fpkm_genes_marker_field_mean))]

# order by DMG identifier
assoc_pgsc_marker_field <- assoc_pgsc_marker_woHK[order(assoc_pgsc_marker_woHK$pgsc_dmg),]

# combine PGSC information with FPKM mean (field)
assoc_pgsc_marker_field <- data.frame(assoc_pgsc_marker_field, 
                                      fpkm_mean_field = fpkm_genes_marker_field_mean)

# order by DMT identifier
assoc_pgsc_marker_field <- assoc_pgsc_marker_field[order(assoc_pgsc_marker_field$pgsc_dmt),]


# only Ct values of check cultivars of field trials
log2_norm_ct_check_field <- subset(log2_norm_ct, 
                                   samplelist$cultivar %in% c("Alegria", "Desiree", "Milva", "Saturna") &
                                   samplelist$cultivation == "field" )
dim(log2_norm_ct_check_field)

log2_norm_ct_check_field_mean <- apply(log2_norm_ct_check_field, 2, mean, na.rm=T)

assoc_pgsc_marker_field = data.frame(assoc_pgsc_marker_field, 
                                     log2_norm_ct_check_field_mean)

plot(assoc_pgsc_marker_field$log2_norm_ct_check_field_mean, 
     log2(assoc_pgsc_marker_field$fpkm_mean_field), pch = 19)
```


## PCA for field trials
```{r PCA for field trials}
# log10 normalized Ct values for all cultivars from field trials
samplelist_field <- droplevels(subset(samplelist, samplelist$cultivation == "field"))

log10_norm_ct_field <- subset(log_norm_ct, samplelist$cultivation == "field" )
dim(log10_norm_ct_field)

log_norm_ct_none_rnipals_field <- func_prep_pca(log10_norm_ct_field, 
                                                scale_method = "none", 
                                                center_option = FALSE, 
                                                pc_number = 5, 
                                                pca_method = "rnipals")
```


### PCA plots for field trials
```{r PCA plots for field trials}
pdf("figures/trost/pca_log_norm_ct_field.pdf")

# treatment
palette(cols_treatment)
func_5pairs_plot(log_norm_ct_none_rnipals_field, samplelist_field, 
                 "treatment", 19, "treatment effect (no scaling, rnipals)")
func_pca_plot(log_norm_ct_none_rnipals_field, 1, 2, samplelist_field, "treatment", 19, 1, 
              "bottomright", 0.8, maintext = "treatment effect (no scaling, rnipals)")

# trial
palette(cols_trial_ordered)
func_5pairs_plot(log_norm_ct_none_rnipals_field, samplelist_field, 
                 "trial", 19, "trial effect (no scaling, rnipals)")

plot(log_norm_ct_none_rnipals_field@scores[,1], 
     log_norm_ct_none_rnipals_field@scores[,2], 
     col = samplelist_field$trial, pch = 19)
legend("topleft", fill = 1:3, 
       c("JKI field 2012", "MPI field 2011", "MPI field 2012"))

plot(log_norm_ct_none_rnipals_field@scores[,3], 
     log_norm_ct_none_rnipals_field@scores[,4], 
     col = samplelist_field$trial, pch = 19, 
     xlim = c(-4.5, 4.5), xlab = "PC3", ylab="PC4", 
     cex.lab = 1.5, cex.axis = 1.2)
legend("bottomright", fill = 1:3, 
       c("JKI field 2012", "MPI field 2011", "MPI field 2012"))

# cultivar
palette(rainbow(34))
func_5pairs_plot(log_norm_ct_none_rnipals_field, samplelist_field, 
                 "cultivar", 19, "cultivar effect (no scaling, rnipals)")

plot(log_norm_ct_none_rnipals_field@scores[,1], 
     log_norm_ct_none_rnipals_field@scores[,3], 
     col = samplelist_field$cultivar, pch = 19)
text(log_norm_ct_none_rnipals_field@scores[,1], 
     log_norm_ct_none_rnipals_field@scores[,3], 
     labels = samplelist_field$cultivar, cex = 0.7)

plot(log_norm_ct_none_rnipals_field@scores[,2], 
     log_norm_ct_none_rnipals_field@scores[,3], 
     col = samplelist_field$cultivar, pch = 19)
text(log_norm_ct_none_rnipals_field@scores[,2], 
     log_norm_ct_none_rnipals_field@scores[,3], 
     labels = samplelist_field$cultivar, cex = 0.7)
dev.off()

palette("default")
```


### Final PCA plots for field trials
```{r final PCA plots for field trials}
log_norm_ct_none_rnipals_field@R2

# trials
pdf("figures/trost/pca_log_norm_ct_trial_field.pdf", width = 6, height = 6)
par(mar=c(4.5, 4.5, 0.5, 0.5))
palette(cols_trial_ordered)
plot(log_norm_ct_none_rnipals_field@scores[,1], 
     log_norm_ct_none_rnipals_field@scores[,5], 
     col = samplelist_field$trial, pch=19, 
     xlab = "PC1 (21.2%)", ylab = "PC5 (5%)", 
     cex.lab = 1.5, cex.axis = 1.2, cex = 1.5)
legend("bottomright", fill = 1:5, 
       c("JKI field 2012", "MPI field 2011", "MPI field 2012"), 
       cex = 1.2, bty = "n")
dev.off()

# treatment
pdf("figures/trost/pca_log_norm_ct_treatment_field.pdf", width = 6, height = 6)
par(mar=c(4.5, 4.5, 0.5, 0.5))
palette(cols_treatment)
plot(log_norm_ct_none_rnipals_field@scores[,1], 
     log_norm_ct_none_rnipals_field@scores[,2], 
     col = samplelist_field$treatment, pch = 19, 
     xlab = "PC1 (21.2%)", ylab = "PC2 (12.9%)", 
     cex.lab = 1.5, cex.axis = 1.2, cex = 1.5)
legend("topright", fill = 1:2, c("control", "drought stress"), 
       cex = 1.2, bty = "n")
dev.off()

# cultivars
pdf("figures/trost/pca_log_norm_ct_cultivar_field.pdf", width=6, height=6)
par(mar=c(4.5, 4.5, 0.5, 0.5))
palette(rainbow(34))
plot(log_norm_ct_none_rnipals_field@scores[,2], 
     log_norm_ct_none_rnipals_field@scores[,3], 
     col = samplelist_field$cultivar, pch=19, 
     xlim = c(-4.5, 4.5), xlab = "PC2 (12.9%)", ylab = "PC3 (8.6%)", 
     cex.lab = 1.5, cex.axis = 1.2, cex = 1.5, ylim = c(-4.5, 4.5))
legend("bottomleft", fill = 1:34, levels(samplelist_field$subspecies_id2), 
       cex = 0.8, horiz = F, ncol = 7, x.intersp = 0.2)
dev.off()
```


## Create heatmap with 6 column types
* Interaction of treatment (2) * trial (3) -> 6

```{r create heatmap with 6 column types}
dim(log_norm_ct)

samplelist_field <- droplevels( samplelist[which(samplelist$cultivation=="field"), ])
samplelist_field_treatment_trial <- interaction(samplelist_field$treatment, samplelist_field$trial)

mat_log_norm_ct <- as.matrix(t(log_norm_ct[which(samplelist$cultivation=="field"), ]))

# colnames(mat_log_norm_ct) 

distance_log_norm_ct = dist(mat_log_norm_ct, method = "euclidian")
cluster_log_norm_ct = hclust(distance_log_norm_ct, method = "average")

distance_log_norm_ct_samples = dist(t(mat_log_norm_ct), method = "euclidian")
cluster_log_norm_ct_samples = hclust(distance_log_norm_ct_samples, method = "average")

#mycluster <- cutree(cluster, h=max(cluster$height)/1.5)
mycluster_tolerance <- cutree(cluster_log_norm_ct, k=5)
#mycluster_tolerance[cluster_log_norm_ct$order]

# 5 colors for different gene clusters
mycolor <- brewer.pal(5, "Set2")
mycolor <- mycolor[as.vector(mycluster_tolerance)]

# get 6 colors for 6 different sample types
mycolor_samples <- brewer.pal(6, "Paired")
mycolor_samples <- mycolor_samples[as.numeric(samplelist_field_treatment_trial)]

my_palette <- colorRampPalette(c("blue", "white", "red"))(n = 299)

col_breaks = c(seq(-2,-0.51,length=100), # for blue
               seq(-0.5,0.5,length=100), # for white
               seq(0.51,2,length=100)) # for red
```


## Plot heatmap with 6 column types
```{r plot heatmap with 6 column types}
pdf("figures/trost/heatmap_qpcr_field.pdf", width=40, height=12)

lmat=rbind(4:3, 2:1)
lhei=c(0.8, 4)
lwid=c(1, 4)

layout(mat = lmat, widths = lwid, heights = lhei)

heatmap.2(mat_log_norm_ct,
  main = "", # heat map title
  density.info = "none",  # turns off density plot inside color legend
  trace = "none",         # turns off trace lines inside the heat map
  margins = c(8,21),     # widens margins around plot
  keysize = 1.8,
  cexRow = 1.5,
  cexCol = 1.5,
  lhei = c(0.6, 4),
  col = my_palette,       # use on color palette defined earlier 
  breaks = col_breaks,    # enable color transition at specified limits
  RowSideColors = mycolor,
  ColSideColors = mycolor_samples,
  Rowv = as.dendrogram(cluster_log_norm_ct),
  Colv = as.dendrogram(cluster_log_norm_ct_samples),
  #labCol = c("GH sensitive", "GH tolerant", "field sensitive", "field tolerant"),
  srtCol = 270,
  adjCol = c(0,1),
  scale = "row",
  dendrogram = "both")     # only draw a row dendrogram

dev.off()
```


## Create heatmap with 34 column types according to cultivars
```{r create heatmap with 34 column types}

mycolor_samples2 <- rainbow(34)
mycolor_samples2 <- mycolor_samples2[as.numeric(samplelist_field$cultivar)]

mat_log_norm_ct2 <- mat_log_norm_ct
colnames(mat_log_norm_ct2) <- samplelist_field$cultivar
```


## Plot heatmap with 34 column types according to cultivars
```{r plot heatmap with 34 column types}
pdf("figures/trost/heatmap_qpcr_field2.pdf", width=40, height=12)

lmat=rbind(4:3, 2:1)
lhei=c(0.8, 4)
lwid=c(1, 4)

layout(mat = lmat, widths = lwid, heights = lhei)

heatmap.2(mat_log_norm_ct,
  main = "", # heat map title
  density.info = "none",  # turns off density plot inside color legend
  trace = "none",         # turns off trace lines inside the heat map
  margins = c(8,21),     # widens margins around plot
  keysize = 1.8,
  cexRow = 1.5,
  cexCol = 1.5,
  lhei = c(0.6, 4),
  col = my_palette,       # use on color palette defined earlier 
  breaks = col_breaks,    # enable color transition at specified limits
  RowSideColors = mycolor,
  ColSideColors = mycolor_samples2,
  Rowv = as.dendrogram(cluster_log_norm_ct),
  Colv = as.dendrogram(cluster_log_norm_ct_samples),
  labCol = samplelist_field$cultivar,
  srtCol = 270,
  adjCol = c(0,1),
  scale = "row",
  dendrogram = "both")     # only draw a row dendrogram

dev.off()
```


# ANOVA with 2 factors
```{r ANOVA with 2 factors}
res_anova_adj <- func_anova_2fac(normalized_values = log_norm_ct, 
                                 trial_factors = samplelist, 
                                 factor1 = "treatment", 
                                 factor2 = "cultivar",
                                 threshold = 0.01,
                                 analyte_names = colnames(log_norm_ct))

head(res_anova_adj)

res_anova_adj_ia <- func_anova_2fac_ia(normalized_values = log_norm_ct, 
                                 trial_factors = samplelist, 
                                 factor1 = "treatment", 
                                 factor2 = "cultivar",
                                 threshold = 0.01,
                                 analyte_names = colnames(log_norm_ct))
```



# Boxplots per gene
```{r boxplots per gene}
pdf("figures/trost/boxplots_per_gene_per_cultivar.pdf", 8, 6)
func_boxplot_1fac(normalized_values = log_norm_ct, 
                  trial_factors = samplelist, 
                  factor = "cultivar", 
                  res_anova_adj = res_anova_adj, 
                  cols = cols_cultivar_34, 
                  analyte_names = colnames(log_norm_ct))
dev.off()

pdf("figures/trost/boxplots_per_gene_per_cultivar_treatment.pdf", 8, 6)
par(mar=c(7, 4.1, 7, 2.1))
func_boxplot_2fac(normalized_values = log_norm_ct, 
                  trial_factors = samplelist, 
                  factor1 = "treatment",
                  factor2 = "cultivar",
                  res_anova_adj = res_anova_adj_ia, 
                  cols = cols_treatment, 
                  analyte_names = colnames(log_norm_ct))
dev.off()
```


# Save workspace & SessionInfo
```{r save workspace}
save.image("qpcr_data_trost.RData")
sessionInfo()
```
