---
title: "Prediction Model using qPCR data (field) from TROST and VALDIS"
author: "Heike Sprenger"
date: "Thursday, March 24, 2016"
output:
  html_document:
    highlight: tango
    number_section: yes
    theme: cerulean
    toc: yes
    toc_depth: 4
---


**Just use common set of 42 genes for TROST and VALDIS dataset**

* 356 TROST samples (5 trials, cultivars)
  * 3 field trials: MPI field 2011, MPI field 2012, JKI field 2012
* 803 VALDIS samples (2 trials, crossings)


# Set working directory  
```{r set working directory}
getwd()
#setwd("D:/work/repos/trost_qpcr/")
```


# Load workspace, packages and scripts
```{r load workspace, message=FALSE}
# load packages
library(knitr)
library(reshape)
library(pander)
library(pcaMethods)
library(ggplot2)
library(corrplot)
library(gplots)
library(RColorBrewer)

library(plyr)
library(rpart)
library(randomForest)
library(partykit)
library(party)
library(caret) 
library(pls)
library(BioMark)
library(varSelRF)
library(psych)
library(corrplot)

# set options for pander
panderOptions('table.split.table', 200)

# set options for knitr
opts_chunk$set(fig.width=5, fig.height=5, cache=FALSE, highlight = TRUE, fig.show="asis")
opts_knit$set(root.dir = '../')

# load workspace
#load("qpcr_model_field_valdis.RData")
```


# Source functions
```{r source functions}
source("../functions/colors.R")
```


# Load data 
```{r load data}
tolerance <- read.table("../trost_phenotypes/output/tolerance_phenotyper_metadata_2sub.txt", 
                        sep = "\t", header = TRUE)

# relevel tolerance factors
tolerance$tol_cat2_fve <- factor(tolerance$tol_cat2_fve, levels = c("low", "high"))
tolerance$tol_cat3_fve <- factor(tolerance$tol_cat3_fve, levels = c("low","mid", "high"))

##############

# TROST data (subset of 42 genes!)
samplelist_trost <- read.table("output/trost/samplelist.txt", header=TRUE, sep="\t")
# --> comes from: "qpcr_data_trost_and_valdis.Rmd"
log_norm_ct_subset_trost <- read.table("output/model/log_norm_ct_trost_subset_42_genes.txt", header=TRUE, sep="\t")
dim(log_norm_ct_subset_trost)
# 356 42

# percentage of NAs
sum(is.na(log_norm_ct_subset_trost)) / (nrow(log_norm_ct_subset_trost)*ncol(log_norm_ct_subset_trost)) *100
# 0.688%

##############

# VALDIS data
samplelist_valdis <- read.table("output/valdis/samplelist.txt", header=TRUE, sep="\t")
log_norm_ct_cleaned_valdis <- read.table("output/valdis/log_norm_ct_cleaned.txt", header=TRUE, sep="\t")
dim(log_norm_ct_cleaned_valdis)
# 803 42

##############

# TROST + VALDIS data --> comes from: "qpcr_data_trost_and_valdis.Rmd"
samplelist_all <- read.table("output/model/samplelist_trost_valdis.txt", header=TRUE, sep="\t")
log_norm_ct_all <- read.table("output/model/log_norm_ct_trost_valdis.txt", header=TRUE, sep="\t")
dim(samplelist_all)
dim(log_norm_ct_all)
# 1159 42
```


## Define TROST training data
```{r define training data}
# join metadata
head(tolerance)
head(samplelist_trost)
levels(samplelist_trost$treatment) <- c("control", "drought stress")

# join TROST samplelist with tolerance information
samplelist_trost_joined <- join(samplelist_trost, tolerance, by="subspecies_id")
samplelist_trost_field <- droplevels( subset(samplelist_trost_joined, samplelist_trost_joined$cultivation=="field"))

# only TROST field data
log_norm_ct_trost_field <- subset(log_norm_ct_subset_trost, samplelist_trost$cultivation=="field")
dim(log_norm_ct_trost_field)

# train/test data
train_data <- subset(log_norm_ct_trost_field, samplelist_trost_field$model_set=="train")
test_data <- subset(log_norm_ct_trost_field, samplelist_trost_field$model_set=="test")

# train/test info
train_info <- subset(samplelist_trost_field, samplelist_trost_field$model_set=="train")
test_info <- subset(samplelist_trost_field, samplelist_trost_field$model_set=="test")
```


## Define TROST control data (and meta data)
```{r define control data}
# data
train_data_control <- subset(train_data, train_info$treatment=="control")
test_data_control <- subset(test_data, test_info$treatment=="control")

# meta data
train_info_control <- subset(train_info, train_info$treatment=="control")
test_info_control <- subset(test_info, test_info$treatment=="control")
```


## Define all TROST data (without 3 cultivars --> model_set=="NA") and subset of all TROST control samples
```{r define all data (control)}
# all TROST samples (without NA --> 3 cultivars)
all_data <- subset(log_norm_ct_trost_field, !samplelist_trost_field$model_set=="NA")
all_info <- subset (samplelist_trost_field, !samplelist_trost_field$model_set=="NA")


# all TROST control samples (without NA --> 3 cultivars)
all_data_control <- subset(log_norm_ct_trost_field, 
                           samplelist_trost_field$treatment=="control" & !samplelist_trost_field$model_set=="NA")

all_info_control <- subset (samplelist_trost_field, 
                            samplelist_trost_field$treatment=="control" & !samplelist_trost_field$model_set=="NA")
```


# Load data without NAs from PCA (rnipals, no scaling, completeObs)
* comes from: ``qpcr_data_trost_and_valdis.Rmd``

```{r data without NAs}
log_norm_ct_woNA_all <- read.table("output/model/log_norm_ct_prep_none_rnipals_completeObs_trost_valdis.txt", header=TRUE, sep="\t")
dim(log_norm_ct_woNA_all)
# 1159 42

#### only TROST field data ####
log_norm_ct_trost_field_woNA <- subset(log_norm_ct_woNA_all, 
                                       samplelist_all$cultivation == "field" & samplelist_all$trost_valdis == "trost")
dim(log_norm_ct_trost_field_woNA)
# 220 42

##### train/test #####
train_data_woNA <- subset(log_norm_ct_trost_field_woNA, samplelist_trost_field$model_set=="train")
test_data_woNA <- subset(log_norm_ct_trost_field_woNA, samplelist_trost_field$model_set=="test")
dim(train_data_woNA)
dim(test_data_woNA)

##### control ####
train_data_control_woNA <- subset(train_data_woNA, train_info$treatment=="control")
test_data_control_woNA <- subset(test_data_woNA, test_info$treatment=="control")

##### all ####
all_data_woNA <- subset(log_norm_ct_trost_field_woNA, !samplelist_trost_field$model_set=="NA")
dim(all_data_woNA)
# 202 42

#### all control samples #####
all_data_control_woNA <- subset(log_norm_ct_trost_field_woNA, 
                                samplelist_trost_field$treatment == "control" & !samplelist_trost_field$model_set == "NA")


#################################################################

#### only VALDIS data #####
log_norm_ct_valdis_woNA <- subset(log_norm_ct_woNA_all, samplelist_all$trost_valdis == "valdis")
dim(log_norm_ct_valdis_woNA)
# 803 42
```


# Random forest regression
## Random forest regression: all data
```{r rf reg all}
input_all_woNA <- data.frame(all_data_woNA, "tol"=all_info$mdrym_fve)

set.seed(1)
rf_all <- randomForest(tol~. , data=input_all_woNA, ntree=1000)
print(rf_all)

#importance(rf_all)
varImpPlot(rf_all)

plot(all_info$mdrym_fve, rf_all$predicted)
abline(0,1)

plot(test_info$mdrym_fve, predict(rf_all, test_data_woNA))
abline(0,1)
sqrt(mean((test_info$mdrym_fve - predict(rf_all, as.data.frame(test_data_woNA)))^2))
```


### Random forest: Predict DRYM (numeric tolerance) from VALDIS data
```{r rf predict DRYM (numeric tolerance) from VALDIS data}
dim(log_norm_ct_valdis_woNA)

rf_all_predicted_drym_valdis <- predict(rf_all, log_norm_ct_valdis_woNA)
summary(rf_all_predicted_drym_valdis)

# change order of genotype levels
samplelist_valdis$crossing <- factor(samplelist_valdis$crossing,
                                     levels = c("Desiree", "Euroresa", "ExA", "Albatros", "AxR", "Ramses"))
```


### Random forest: Sort and aggregate DRYM values
```{r sort and aggregate drym values}
# join predicted DRYM values with line names
rf_all_predicted_drym_valdis_line <- data.frame(drym = rf_all_predicted_drym_valdis, 
                                                line = samplelist_valdis$name)
head(rf_all_predicted_drym_valdis_line)

# aggregate median and sort
rf_all_predicted_drym_valdis_median <- aggregate(rf_all_predicted_drym_valdis_line$drym, 
                                                 by=list(rf_all_predicted_drym_valdis_line$line),
                                                 median)
colnames(rf_all_predicted_drym_valdis_median) <- c("line", "drym")
rf_all_predicted_drym_valdis_median_sorted <- rf_all_predicted_drym_valdis_median[order(rf_all_predicted_drym_valdis_median$drym),]


# save median of predicted drym
write.table(rf_all_predicted_drym_valdis_median, 
            "output/model/rf_all_predicted_drym_valdis_median.txt", 
            sep="\t", row.names=F)


# sort predicted DRYM values for plot
rf_all_predicted_drym_valdis_sorted <- with(rf_all_predicted_drym_valdis_line, reorder(line, drym, median, na.rm=T))
head(rf_all_predicted_drym_valdis_sorted)
```


### Random forest: Top35
```{r top35}
sens35 <- head(rf_all_predicted_drym_valdis_median_sorted$line, n=35)
tol35 <- tail(rf_all_predicted_drym_valdis_median_sorted$line, n=35)

sens35_idx <- which(rf_all_predicted_drym_valdis_line$line %in% sens35)
tol35_idx <- which(rf_all_predicted_drym_valdis_line$line %in% tol35)

rf_all_predicted_drym_valdis_line_sens35 <- rf_all_predicted_drym_valdis_line[sens35_idx,]
rf_all_predicted_drym_valdis_line_tol35 <- rf_all_predicted_drym_valdis_line[tol35_idx,]

# sort
rf_all_predicted_drym_valdis_line_sens35_sorted <- with(rf_all_predicted_drym_valdis_line_sens35, reorder(line, drym, median, na.rm=T))
rf_all_predicted_drym_valdis_line_sens35_sorted <- droplevels(rf_all_predicted_drym_valdis_line_sens35_sorted)
rf_all_predicted_drym_valdis_line_tol35_sorted <- with(rf_all_predicted_drym_valdis_line_tol35, reorder(line, drym, median, na.rm=T))
rf_all_predicted_drym_valdis_line_tol35_sorted <- droplevels(rf_all_predicted_drym_valdis_line_tol35_sorted)
```


### Random forest: Define colors
```{r define colors}
color <- rep("gold", 199)
color [which( grepl("^AxR", levels(rf_all_predicted_drym_valdis_sorted) ))] <- "greenyellow"
color [which( levels (rf_all_predicted_drym_valdis_sorted) %in% c("Ramses", "Euroresa", "Albatros") )] <- "red"
color [which( levels (rf_all_predicted_drym_valdis_sorted) == "Desiree" )] <- "deepskyblue"
color_fac <- factor(color)

color_sens35 <- rep("gold", 35)
color_sens35 [which( grepl("^AxR", levels(rf_all_predicted_drym_valdis_line_sens35_sorted) ))] <- "greenyellow"
color_sens35 [which( levels (rf_all_predicted_drym_valdis_line_sens35_sorted) %in% c("Ramses", "Euroresa", "Albatros") )] <- "red"
color_sens35 [which( levels (rf_all_predicted_drym_valdis_line_sens35_sorted) == "Desiree" )] <- "deepskyblue"

color_tol35 <- rep("gold", 35)
color_tol35 [which( grepl("^AxR", levels(rf_all_predicted_drym_valdis_line_tol35_sorted) ))] <- "greenyellow"
color_tol35 [which( levels (rf_all_predicted_drym_valdis_line_tol35_sorted) %in% c("Ramses", "Euroresa", "Albatros") )] <- "red"
color_tol35 [which( levels (rf_all_predicted_drym_valdis_line_tol35_sorted) == "Desiree" )] <- "deepskyblue"
```


### Random forest: Boxplot predicted DRYM
```{r boxplot predicted drym}
pdf("figures/model/valdis_boxplot_rf_all_predicted_drym.pdf", width=8.5, height=5)
boxplot(rf_all_predicted_drym_valdis ~ samplelist_valdis$crossing, ylab="DRYM", cex.lab=1.5, cex.axis=1.1)
dev.off()


pdf("figures/model/valdis_boxplot_rf_all_predicted_drym_line.pdf", width=12, height=6)
boxplot(rf_all_predicted_drym_valdis ~ samplelist_valdis$name, 
        ylab="DRYM", cex.lab=1.5, cex.axis=0.7, las=2)
boxplot(drym ~ rf_all_predicted_drym_valdis_sorted, data = rf_all_predicted_drym_valdis_line, 
        ylab="DRYM", cex.lab=1.5, cex.axis=0.7, las=2, col=color)
legend("bottomright", fill=levels(color_fac), legend=c("Desiree", "ExA", "AxR", "Eltern"))
dev.off()

# SENS 35
pdf("figures/model/valdis_boxplot_rf_all_predicted_drym_line_sens35.pdf", width=10, height=5)
par(mar=c(7,5,1,1))
boxplot(drym ~ rf_all_predicted_drym_valdis_line_sens35_sorted, data = rf_all_predicted_drym_valdis_line_sens35, 
        ylab="DRYM", cex.lab=1.5, cex.axis=1.2, las=2, col=color_sens35)
legend("bottomright", fill=levels(color_fac), legend=c("Desiree", "ExA", "AxR", "Eltern"), cex=0.9)
dev.off()

# TOL 35
pdf("figures/model/valdis_boxplot_rf_all_predicted_drym_line_tol35.pdf", width=10, height=5)
par(mar=c(7,5,1,1))
boxplot(drym ~ rf_all_predicted_drym_valdis_line_tol35_sorted, data = rf_all_predicted_drym_valdis_line_tol35, 
        ylab="DRYM", cex.lab=1.5, cex.axis=1.2, las=2, col=color_tol35)
legend("bottomright", fill=levels(color_fac), legend=c("Desiree", "ExA", "AxR", "Eltern"), cex=0.9)
dev.off()
```


### Cross-Validation
```{r rf reg all CV}
set.seed(3)
rf_all_cv <- rfcv(all_data_woNA, all_info$mdrym_fve, step=0.8)
rf_all_cv$n.var
with(rf_all_cv, plot(n.var, error.cv, log="x", type="o", lwd=2))
```


# Lasso model by glmnet ALL data
## Lasso: Define data
```{r lasso define data}
par(mar=c(5, 5, 1, 1))

# TEST data
test_x <- as.matrix(test_data_woNA)
test_y <- test_info$mdrym_fve

# use TRAIN data
train_x <- as.matrix(train_data_woNA)
train_y <- train_info$mdrym_fve

# use ALL data
x <- as.matrix(all_data_woNA)
y <- all_info$mdrym_fve
```


## Lasso: Model with TRAINING data
```{r lasso train data}
set.seed(1)
lasso_train <- glmnet(train_x, train_y)
#plot(lasso_train)

# Cross-validation
set.seed(1)
lasso_cv_train <- cv.glmnet(train_x, train_y)
plot(lasso_cv_train, ylim=c(0, 0.005))

# use no specific lambda
lasso_train_pred_test <- predict(lasso_train, new = test_x)

# calculate mean error
lasso_train_pred_test_err <- apply( (lasso_train_pred_test - test_y)^2, 2, mean )
points(log(lasso_train$lambda), lasso_train_pred_test_err, col="blue", pch="*")

# use minimal lamba value
lasso_train_pred_train_lambda <- predict(lasso_cv_train, new = train_x, s="lambda.min")
lasso_train_pred_test_lambda <- predict(lasso_cv_train, new = test_x, s="lambda.min")

dim(lasso_train_pred_train_lambda)
dim(lasso_train_pred_test_lambda)

plot(train_y, lasso_train_pred_train_lambda[,1])
cor.test(train_y, lasso_train_pred_train_lambda[,1])

plot(test_y, lasso_train_pred_test_lambda[,1])
cor.test(test_y, lasso_train_pred_test_lambda[,1])
```


## Lasso: Model with ALL data
```{r lasso ALL data}
set.seed(1)
lasso_all <- glmnet(x, y)
#plot(lasso_all)

# Cross-validation
set.seed(1)
lasso_cv_all <- cv.glmnet(x, y)
plot(lasso_cv_all, ylim=c(0, 0.005))

# use no specific lambda
lasso_all_pred_test <- predict(lasso_all, new = test_x)

# calculate mean error
lasso_all_pred_test_err <- apply( (lasso_all_pred_test - test_y)^2, 2, mean )
points(log(lasso_all$lambda), lasso_all_pred_test_err, col="blue", pch="*")

# use optimal lamba value
lasso_all_pred_all_lambda <- predict(lasso_cv_all, new = x, s="lambda.min")
lasso_all_pred_test_lambda <- predict(lasso_cv_all, new = test_x, s="lambda.min")
```


## Lasso: Compare models based on training or all data
```{r lasso compare models based on training or all data}
plot(lasso_train_pred_test_lambda, lasso_all_pred_test_lambda)
plot(test_y, lasso_all_pred_test_lambda)
```


## Lasso: Plot CV mean error vs. lambda and number of non-zero coef vs. lambda
```{r lasso plot CV mean error vs. lambda and number of non-zero coef vs. lambda}

pdf("figures/model/valdis_lasso_cv_all_lambda_2.pdf")

par(mar=c(5,5,3,5))

plot(lasso_cv_all$lambda, lasso_cv_all$cvm, type="l", lwd=2, 
     xlab="lambda", ylab="CV mean error", cex.lab=1.5, 
     main="transcript lasso model: CV mean error vs. lambda \n and number of variables vs. lambda")
grid(NA, 8, lwd = 2) # grid only in y-direction
abline(v=lasso_cv_all$lambda.1se, lty=2, col="green", lwd=2)
text(0.004, 0.0013, "lambda.1se", col="green", cex=1.3)

par(new=TRUE)

plot(lasso_cv_all$lambda, lasso_cv_all$nzero, type="l", 
     lwd=2 ,col="blue",xaxt="n",yaxt="n",xlab="",ylab="")
axis(4, col="blue", col.ticks="blue") # add second y-axis
mtext("number of variables",side=4,line=3, col="blue", cex=1.5)

dev.off()
```


## Lasso: Extract coefficients at a single value of lambda
```{r lasso extract coefficients at a single value of lambda}
coef(lasso_all,s=0.01)
coef(lasso_all,s=0.001)
coef(lasso_all,s=0.0001)
# smaller value for s --> more coefficients kept in model
# higher value for s --> less coefficients kept in model

lasso_cv_all_coef <- predict(lasso_cv_all, type="coefficients", s=lasso_cv_all$lambda.min)
write.table(as.matrix(lasso_cv_all_coef), "output/model/lasso_cv_all_coef.txt", sep="\t")

lasso_cv_all_coef_1se <- predict(lasso_cv_all, type="coefficients", s=lasso_cv_all$lambda.1se)
write.table(as.matrix(lasso_cv_all_coef_1se), "output/model/lasso_cv_all_coef_1se.txt", sep="\t")

# identical result:
coef(lasso_cv_all , s = "lambda.min") # 5 variables (of 42) have coef of zero --> 37 left
coef(lasso_cv_all , s = "lambda.1se") # 19 variables (of 42) have coef of zero --> 23 left

# plot RF vs LASSO coefficients/importance
plot(importance(rf_all), abs(as.matrix(coef(lasso_cv_all , s = 1.838797e-05))[-1,]))
plot(abs(as.matrix(coef(lasso_cv_all , s = 1.838797e-05))))
varImpPlot(rf_all)
```


## Lasso: Plot observed vs. predicted DRYM values
```{r lasso plot observed vs. predicted DRYM values}
par(mfrow=c(1,2))

# predict test data
plot(test_y, lasso_all_pred_test_lambda, pch=19, yaxt="n", xaxt="n", main="lasso", xlim=c(-0.1, 0.1), ylim=c(-0.1, 0.1))
axis(1, at=seq(-0.1, 0.1, 0.01))
axis(2, at=seq(-0.1, 0.1, 0.01))
abline(0,1, col="red")


# plot for random forest model
plot(test_info$mdrym_fve, predict(rf_all, test_data_woNA), pch=19, yaxt="n", xaxt="n", main="random forest", xlim=c(-0.1, 0.1), ylim=c(-0.1, 0.1))
axis(1, at=seq(-0.1, 0.1, 0.01))
axis(2, at=seq(-0.1, 0.1, 0.01))
abline(0,1, col="red")


# predict training data (here: all data were used for training)
plot(y, lasso_all_pred_all_lambda, pch=19, yaxt="n", xaxt="n", main="lasso", xlim=c(-0.1, 0.1), ylim=c(-0.1, 0.1))
axis(1, at=seq(-0.1, 0.1, 0.01))
axis(2, at=seq(-0.1, 0.1, 0.01))
abline(0,1, col="red")

# plot for random forest model
plot(all_info$mdrym_fve, rf_all$predicted, pch=19, yaxt="n", xaxt="n", main="random forest", xlim=c(-0.1, 0.1), ylim=c(-0.1, 0.1))
axis(1, at=seq(-0.1, 0.1, 0.01))
axis(2, at=seq(-0.1, 0.1, 0.01))
abline(0,1, col="red")

par(mfrow=c(1,1))
```


## Lasso: Predict DRYM (numeric tolerance) from VALDIS data
```{r lasso predict DRYM (numeric tolerance) from VALDIS data}
#log_norm_ct_valdis_woNA

# use minimal lamba value
lasso_all_predicted_drym_valdis <- predict(lasso_cv_all, new = as.matrix(log_norm_ct_valdis_woNA), s="lambda.min")

# use largest value of lambda such that CV-error is within 1 standard error of the minimum (ca. 50 metabolites)
lasso_all_predicted_drym_valdis_1se <- predict(lasso_cv_all, new = as.matrix(log_norm_ct_valdis_woNA), s="lambda.1se")

summary(lasso_all_predicted_drym_valdis)
summary(rf_all_predicted_drym_valdis)

hist(lasso_all_predicted_drym_valdis[,1])
hist(rf_all_predicted_drym_valdis)

# random forest vs. lasso prediction of DRYM
plot(lasso_all_predicted_drym_valdis[,1], rf_all_predicted_drym_valdis)
plot(lasso_all_predicted_drym_valdis_1se[,1], rf_all_predicted_drym_valdis)

pdf("figures/model/valdis_boxplot_compare_rf_lasso_predicted_drym.pdf", 5,5)
boxplot(rf_all_predicted_drym_valdis, lasso_all_predicted_drym_valdis[,1], names=c("Random Forest", "LASSO"), 
        ylab="DRYM", cex.axis=1.2, cex.lab=1.4, main="transcript model")
dev.off()

# plot model with minimal CV-error (37 transcripts) vs. model with CV-error within 1SE (23 transcripts)
plot(lasso_all_predicted_drym_valdis[,1], lasso_all_predicted_drym_valdis_1se[,1],
     xlim=c(-0.1, 0.12), ylim=c(-0.1, 0.12))
abline(0,1, col="red")
```


## Lasso: Sort and aggregate DRYM values
```{r lasso sort and aggregate drym values}
# remove DESIREE
lasso_all_predicted_drym_valdis_without_desiree <- subset(lasso_all_predicted_drym_valdis,
                                            samplelist_valdis$name != "Desiree")

samplelist_valdis_without_desiree <- subset(samplelist_valdis,
                                            samplelist_valdis$name != "Desiree")

samplelist_valdis_without_desiree <- droplevels(samplelist_valdis_without_desiree)

# join predicted DRYM values with line names
lasso_all_predicted_drym_valdis_line <- data.frame(drym = lasso_all_predicted_drym_valdis_without_desiree[,1], 
                                                line = samplelist_valdis_without_desiree$name)
head(lasso_all_predicted_drym_valdis_line)

# aggregate median and sort
lasso_all_predicted_drym_valdis_median <- aggregate(lasso_all_predicted_drym_valdis_line$drym, 
                                                 by=list(lasso_all_predicted_drym_valdis_line$line),
                                                 median)
colnames(lasso_all_predicted_drym_valdis_median) <- c("line", "drym")
lasso_all_predicted_drym_valdis_median_sorted <- lasso_all_predicted_drym_valdis_median[order(lasso_all_predicted_drym_valdis_median$drym),]

# save median of predicted drym
head(lasso_all_predicted_drym_valdis_median)
write.table(lasso_all_predicted_drym_valdis_median, "output/model/lasso_all_predicted_drym_valdis_median.txt", sep="\t", row.names=F)
write.table(lasso_all_predicted_drym_valdis_line, "output/valdis/lasso_all_predicted_drym_valdis.txt", sep="\t", row.names=F)

# sort predicted DRYM values for plot
lasso_all_predicted_drym_valdis_sorted <- with(lasso_all_predicted_drym_valdis_line, reorder(line, drym, median, na.rm=T))
head(lasso_all_predicted_drym_valdis_sorted)
```


## Lasso: Sort and aggregate DRYM values (1SE model)
```{r lasso sort and aggregate drym values (1SE model)}
# join predicted DRYM values with line names
lasso_all_predicted_drym_valdis_line_1se <- data.frame(drym = lasso_all_predicted_drym_valdis_1se[,1], 
                                                line = samplelist_valdis$name)

# aggregate median 
lasso_all_predicted_drym_valdis_median_1se <- aggregate(lasso_all_predicted_drym_valdis_line_1se$drym, 
                                                 by=list(lasso_all_predicted_drym_valdis_line_1se$line),
                                                 median)
colnames(lasso_all_predicted_drym_valdis_median_1se) <- c("line", "drym")

# save median of predicted drym
write.table(lasso_all_predicted_drym_valdis_median_1se, 
            "output/model/lasso_all_predicted_drym_valdis_median_1se.txt", 
            sep="\t", row.names=F)
```


## Lasso: Top35
```{r lasso top35}
lasso_sens35 <- head(lasso_all_predicted_drym_valdis_median_sorted$line, n=35)
lasso_tol35 <- tail(lasso_all_predicted_drym_valdis_median_sorted$line, n=35)

lasso_sens35_idx <- which(lasso_all_predicted_drym_valdis_line$line %in% lasso_sens35)
lasso_tol35_idx <- which(lasso_all_predicted_drym_valdis_line$line %in% lasso_tol35)

lasso_all_predicted_drym_valdis_line_sens35 <- lasso_all_predicted_drym_valdis_line[lasso_sens35_idx,]
lasso_all_predicted_drym_valdis_line_tol35 <- lasso_all_predicted_drym_valdis_line[lasso_tol35_idx,]

# sort
lasso_all_predicted_drym_valdis_line_sens35_sorted <- with(lasso_all_predicted_drym_valdis_line_sens35, 
                                                           reorder(line, drym, median, na.rm=T))
lasso_all_predicted_drym_valdis_line_sens35_sorted <- droplevels(lasso_all_predicted_drym_valdis_line_sens35_sorted)
lasso_all_predicted_drym_valdis_line_tol35_sorted <- with(lasso_all_predicted_drym_valdis_line_tol35, 
                                                          reorder(line, drym, median, na.rm=T))
lasso_all_predicted_drym_valdis_line_tol35_sorted <- droplevels(lasso_all_predicted_drym_valdis_line_tol35_sorted)
```


## Lasso: Define colors
```{r lasso define colors}
lasso_color <- rep("#00756D", 198) # for ExA
lasso_color [which( grepl("^AxR", levels(lasso_all_predicted_drym_valdis_sorted) ))] <- "#BF5300"
lasso_color [which( levels (lasso_all_predicted_drym_valdis_sorted) == "Albatros" )] <- "grey"
lasso_color [which( levels (lasso_all_predicted_drym_valdis_sorted) == "Euroresa" )] <- "#5778B9"
lasso_color [which( levels (lasso_all_predicted_drym_valdis_sorted) == "Ramses" )] <- "#F7B944"
lasso_color_fac <- factor(lasso_color)

lasso_color_sens35 <- rep("#00756D", 35)
lasso_color_sens35 [which( grepl("^AxR", levels(lasso_all_predicted_drym_valdis_line_sens35_sorted) ))] <- "#BF5300"
lasso_color_sens35 [which( levels (lasso_all_predicted_drym_valdis_line_sens35_sorted) == "Albatros" )] <- "grey"
lasso_color_sens35 [which( levels (lasso_all_predicted_drym_valdis_line_sens35_sorted) == "Euroresa" )] <- "#5778B9"
lasso_color_sens35 [which( levels (lasso_all_predicted_drym_valdis_line_sens35_sorted) == "Ramses" )] <- "#F7B944"

lasso_color_tol35 <- rep("#00756D", 35)
lasso_color_tol35 [which( grepl("^AxR", levels(lasso_all_predicted_drym_valdis_line_tol35_sorted) ))] <- "#BF5300"
lasso_color_tol35 [which( levels (lasso_all_predicted_drym_valdis_line_tol35_sorted) == "Albatros" )] <- "grey"
lasso_color_tol35 [which( levels (lasso_all_predicted_drym_valdis_line_tol35_sorted) == "Euroresa" )] <- "#5778B9"
lasso_color_tol35 [which( levels (lasso_all_predicted_drym_valdis_line_tol35_sorted) == "Ramses" )] <- "#F7B944"

```


## Lasso: Boxplot predicted DRYM
```{r lasso boxplot predicted drym}
pdf("figures/model/valdis_boxplot_lasso_all_predicted_drym.pdf", width=6, height=5)
par(mar=c(3,5,2,1))
boxplot(lasso_all_predicted_drym_valdis_without_desiree ~ samplelist_valdis_without_desiree$crossing, ylab="DRYM", 
        cex.lab=1.5, cex.axis=1.1, main = "transcript model", col=cols_genotype_report[c(3,4,1,2,5)], 
        ylim=c(-0.2, 0.15), yaxt="n")
axis(side=2, labels=c("-0.2", "-0.1", "0", "0.1"), at=c(-0.2, -0.1, 0, 0.1), cex.axis=1.1)
dev.off()


pdf("figures/model/valdis_boxplot_lasso_all_predicted_drym_line.pdf", width=12, height=6)
par(mar=c(4.5,5,2,0.5))
boxplot(lasso_all_predicted_drym_valdis ~ samplelist_valdis$name, 
        ylab="DRYM", cex.lab=1.5, cex.axis=0.7, las=2, main = "transcript model")
boxplot(drym ~ lasso_all_predicted_drym_valdis_sorted, data = lasso_all_predicted_drym_valdis_line, 
        ylab="DRYM", cex.lab=1.5, cex.axis=0.7, las=2, col=lasso_color, main = "transcript model", names=F, xlab="genotype")
legend("bottomright", fill=levels(lasso_color_fac), legend=c("ExA", "Euroresa", "AxR", "Ramses", "Albatros"), horiz=T)
dev.off()

# SENS 35
pdf("figures/model/valdis_boxplot_lasso_all_predicted_drym_line_sens35.pdf", width=10, height=5)
par(mar=c(7,5,1,1))
boxplot(drym ~ lasso_all_predicted_drym_valdis_line_sens35_sorted, data = lasso_all_predicted_drym_valdis_line_sens35, 
        ylab="DRYM", cex.lab=1.5, cex.axis=1.2, las=2, col=lasso_color_sens35, main = "transcript model: sens35")
legend("bottomright", fill=levels(lasso_color_fac), legend=c("ExA", "Euroresa", "AxR", "Ramses", "Albatros"), cex=0.9)
dev.off()

# TOL 35
pdf("figures/model/valdis_boxplot_lasso_all_predicted_drym_line_tol35.pdf", width=10, height=5)
par(mar=c(7,5,1,1))
boxplot(drym ~ lasso_all_predicted_drym_valdis_line_tol35_sorted, data = lasso_all_predicted_drym_valdis_line_tol35, 
        ylab="DRYM", cex.lab=1.5, cex.axis=1.2, las=2, col=lasso_color_tol35, main = "transcript model: tol35")
legend("bottomright", fill=levels(lasso_color_fac), legend=c("ExA", "Euroresa", "AxR", "Ramses", "Albatros"), cex=0.9)
dev.off()
```


# Save workspace and sessionInfo
```{r save workspace}
save.image("qpcr_model_field_valdis.RData")
sessionInfo()
```

