---
title: "Analysis of qRT-PCR data from TROST breeder trials"
author: "Heike Sprenger"
date: "Thursday, March 17, 2016"
output:
  html_document:
    highlight: tango
    number_section: yes
    theme: cerulean
    toc: yes
    toc_depth: 4
---

# Set working directory  
```{r set working directory}
getwd()
#setwd("D:/work/repos/trost_qpcr/")
```


# Load workspace, packages and scripts
```{r load workspace, message=FALSE}
# load packages
library(knitr)
library(reshape)
library(pander)
library(pcaMethods)
library(ggplot2)
library(corrplot)
library(gplots)
library(RColorBrewer)

# set options for pander
panderOptions('table.split.table', 200)

# set options for knitr
opts_chunk$set(fig.width=5, fig.height=5, cache=FALSE, highlight = TRUE, fig.show="asis")
opts_knit$set(root.dir = '../')

# load workspace
#load("qpcr_data_breeder.RData")
```


# Source R functions
```{r source R functions}
source("../functions/colors.R")
source("../functions/func_prep_pca.R")
source("../functions/func_pca_plots.R")
source("../functions/RemoveFactors_function.R")
source("../functions/func_anova.R")
source("../functions/func_boxplot.R")
```


# Load sample information
```{r load sample information}
samplelist <- read.table("input/breeder/samplelist_breeder.txt", header=TRUE, sep="\t")

# remove sample with many high Ct values and Undetermined values
which(samplelist$sampleID == "885463")
samplelist <- samplelist[-200,]

samplelist$year <- as.factor(samplelist$year)

write.table(samplelist, "output/breeder/samplelist.txt", sep="\t")
```


# Load primer info
```{r load primer info}
primer_info <- read.table("input/breeder/primer_info_breeder.txt", header=TRUE, sep="\t")
```


# Load functional annotation
```{r load functional annotation}
assoc_pgsc <- read.table("../trost_transcriptomics/data/PGSC_DM_v3.4_g2t2c2p2func_edit.txt", sep="\t")
colnames(assoc_pgsc) <- c("pgsc_dmg", "pgsc_dmt", "pgsc_dmc", "pgsc_dmp", "func")
head(assoc_pgsc)
```


# Load qpcr raw data (after cleaning) and modify them
```{r load data}
rawdata <- read.table("input/breeder/results_final_breeder.txt", header = T, sep="\t", check.names = F)

# remove sample with many high Ct values and Undetermined values
remove_idx <- which(rawdata$sampleID == "885463")
rawdata <- rawdata[-remove_idx,]

# replace Undetermined Ct values by NA (or 40???)
undetermined_idx <- which(rawdata$Ct == "Undetermined")
rawdata$Ct <- as.numeric(as.character(rawdata$Ct))
rawdata$Ct[undetermined_idx] <- NA

sum(is.na(rawdata$Ct))

# replace problematic Ct values by NA
peak_ok_idx <- which(rawdata$peak_comment == "ok")
rawdata_cleaned <- rawdata
rawdata_cleaned$Ct[-peak_ok_idx] <- NA
hist(rawdata_cleaned$Ct)

sum(is.na(rawdata_cleaned$Ct))

table(rawdata_cleaned$primer_name)

# order by DMT id
rawdata_cleaned <- rawdata_cleaned[order(rawdata_cleaned$sampleID), ]

# merge with primer info
rawdata_cleaned_merge <- merge(rawdata_cleaned, primer_info, by = "primer_name")

# extract table with sampleID, DMT, and Ct value
rawdata_cleaned_values <- rawdata_cleaned_merge[,c(3,14,10)]
head(rawdata_cleaned_values)

# sampleID in rows, DMT in columns
rawdata_cleaned_values_cast <- cast(rawdata_cleaned_values, sampleID ~ DMT, value = "Ct")

rownames(rawdata_cleaned_values_cast) <- rawdata_cleaned_values_cast$sampleID
rawdata_cleaned_ct <- as.data.frame(rawdata_cleaned_values_cast[,-1])
dim(rawdata_cleaned_ct)
# rows: 203 samples
# columns: 47 genes

# columns with HK genes
HK_index <- which(colnames(rawdata_cleaned_ct) %in% c("PGSC0003DMT400030611", 
                                                      "PGSC0003DMT400068116", 
                                                      "PGSC0003DMT400037585", 
                                                      "PGSC0003DMT400080569"))

# values without HK genes
rawdata_cleaned_ct_woHK <-rawdata_cleaned_ct[,-HK_index]
dim(rawdata_cleaned_ct_woHK)
# 203 samples, 43 genes

# values of HK genes
rawdata_cleaned_ct_HK <- rawdata_cleaned_ct[,HK_index]

# heatmap
# heatmap.2(as.matrix(rawdata_cleaned_ct_HK),
#           labCol = colnames(rawdata_cleaned_ct_HK),
#           srtCol = 45, cexCol = 1)

#write.table(colnames(rawdata_cleaned_ct), "../output/breeder/PGSC_DMT_identifier_qpcr_marker_HK.txt", sep="\t")
```


# Plot ct values
```{r plot ct values}
# boxplot of all ct-values vs. genes, first 4 are HG genes
cols_47_genes <- rep("lightblue", 47)
cols_47_genes[HK_index] <- "red"
boxplot.matrix(as.matrix(rawdata_cleaned_ct), 
               use.cols = TRUE, col = cols_47_genes, 
               las = 2, cex.lab = 0.8)

# plot of HK gene 1 vs. HK gene 2
plot(rawdata_cleaned_ct[,HK_index[1]], rawdata_cleaned_ct[,HK_index[2]])

#class(rawdata_cleaned_ct)
p <- ggplot(as.data.frame(rawdata_cleaned_ct), aes(PGSC0003DMT400030611, PGSC0003DMT400080569))
p + geom_point(alpha =1/3, size=5) 

# histogram of ct-values per gene
pdf("figures/breeder/hist_rawdata_cleaned_ct.pdf")
for (i in 1:43){
hist(rawdata_cleaned_ct_woHK[,i], breaks=30, col="grey", main=colnames(rawdata_cleaned_ct_woHK)[i])
}
dev.off()

# pairs plot of first 10 genes
pairs(rawdata_cleaned_ct_woHK[,1:10])
```


# Calculate mean of 4 housekeeping genes per sample
```{r calculate mean of 4 housekeeping genes per sample}
housekeeping_mean_all <- apply(rawdata_cleaned_ct_HK, 1, mean, na.rm=TRUE)

# without HK4 !
# HK_index_part <- which(colnames(rawdata_cleaned_ct) %in% c("PGSC0003DMT400068117", 
#                                      "PGSC0003DMT400037585", 
#                                      "PGSC0003DMT400080569"))
# housekeeping_mean <- apply(rawdata_cleaned_ct[,HK_index_part], 1, mean)
```


# Calculate delta Ct: ``CT_gene - mean(CT_HK)``
```{r calculate delta Ct}
delta_ct <- matrix(1, nrow=nrow(rawdata_cleaned_ct_woHK), ncol=ncol(rawdata_cleaned_ct_woHK))
colnames(delta_ct) <- colnames(rawdata_cleaned_ct_woHK)
rownames(delta_ct) <- rownames(rawdata_cleaned_ct_woHK)

for (i in 1:nrow(rawdata_cleaned_ct_woHK)){
  for (j in 1:ncol(rawdata_cleaned_ct_woHK)){
    delta_ct[i,j] <- rawdata_cleaned_ct_woHK[i,j]-housekeeping_mean_all[i]
    }
  }
```


# Calculate 2 to the power of minus delta Ct
```{r calculate 2 to the power of minus delta Ct}
two_minus_delta_ct <- 2^-(delta_ct)
hist(two_minus_delta_ct, breaks=50, col="grey")

#apply(two_minus_delta_ct, 2, max, na.rm=T)

# log 10 transformation of normalized values (2 to the power of minus delta Ct)
log_norm_ct <- log10(two_minus_delta_ct)
hist(log_norm_ct, col="grey", breaks=50)
# low values correpond to low expression in comparison to HK genes (e.g. CT 40)
# high values correpond to high expression in comparison to HK genes (e.g. CT 20)

# log 2 transformation
log2_norm_ct <- log2(two_minus_delta_ct)
log2_norm_ct[1:5,1:5]
two_minus_delta_ct[1:5,1:5]

# export expression data
write.table(two_minus_delta_ct, "output/breeder/two_minus_delta_ct.txt", sep="\t")
write.table(log2_norm_ct, "output/breeder/log2_norm_ct.txt", sep="\t")
write.table(log_norm_ct, "output/breeder/log_norm_ct.txt", sep="\t")
```


# Heatmap of ``two_minus_delta_ct`` and ``log2_norm_ct``
```{r heatmap of two_minus_delta_ct and log2_norm_ct}
my_palette <- colorRampPalette(c("white", "red"))(n = 199)
col_breaks <- c(seq(0, 3, length=100), 
                seq(3.1, 100, length=100))

heatmap.2(as.matrix(two_minus_delta_ct), trace = "none",
          labCol = colnames(two_minus_delta_ct),
          labRow = rownames(two_minus_delta_ct),
          col = my_palette, breaks = col_breaks)



my_palette <- colorRampPalette(c("blue","white", "red"))(n = 299)
col_breaks <- c(seq(-20, -3.1, length=100), 
                seq(-3, 3, length=100),
                seq(3.1, 20, length=100))

heatmap.2(as.matrix(log2_norm_ct), trace = "none",
          labCol = colnames(log2_norm_ct),
          labRow = rownames(log2_norm_ct),
          col = my_palette, breaks = col_breaks)
```


# PCA
```{r PCA}
log_norm_ct_pareto_rnipals <- func_prep_pca(log_norm_ct, scale_method = "pareto", center_option = FALSE, 
                                             pc_number = 5, pca_method = "rnipals")

log_norm_ct_none_rnipals <- func_prep_pca(log_norm_ct, scale_method = "none", center_option = FALSE, 
                                             pc_number = 5, pca_method = "rnipals")

log_norm_ct_none_ppca <- func_prep_pca(log_norm_ct, scale_method = "none", center_option = FALSE, 
                                             pc_number = 5, pca_method = "ppca")


# export completeObs for later use by RandomForest
#write.table(log_norm_ct_none_rnipals@completeObs, "output/breeder/log_norm_ct_none_rnipals_completeObs.txt", sep="\t")
```


## PCA plots
```{r PCA plots}
pdf("figures/breeder/pca_log_norm_ct.pdf")
# no scaling, rnipals
# year
palette(heike_palette_2)
func_5pairs_plot(log_norm_ct_none_rnipals, samplelist, "year", 19, "year effect (no scaling, rnipals)")
func_pca_plot(log_norm_ct_none_rnipals, 1, 2, samplelist, "year", 19, 1, 
              "bottomright", 0.8, maintext="year effect (no scaling, rnipals)")

# location
palette(heike_palette_3)
func_5pairs_plot(log_norm_ct_none_rnipals, samplelist, "location_name", 19, "location effect (no scaling, rnipals)")
func_pca_plot(log_norm_ct_none_rnipals, 1, 2, samplelist, "location_name", 19, 1, 
              "bottomright", 0.8, maintext="location effect (no scaling, rnipals)")

# trial
palette(brewer.pal(6, "Paired"))
func_5pairs_plot(log_norm_ct_none_rnipals, samplelist, "trial", 19, "trial effect (no scaling, rnipals)")
func_pca_plot(log_norm_ct_none_rnipals, 1, 2, samplelist, "trial", 19, 1, 
              "bottomright", 0.8, maintext="trial effect (no scaling, rnipals)")

# cultivar
palette(rainbow(34))
func_5pairs_plot(log_norm_ct_none_rnipals, samplelist, "cultivar", 19, "cultivar effect (no scaling, rnipals)")

plot(log_norm_ct_none_rnipals@scores[,1], log_norm_ct_none_rnipals@scores[,3], col=samplelist$cultivar, pch=19)
text(log_norm_ct_none_rnipals@scores[,1], log_norm_ct_none_rnipals@scores[,3], labels=samplelist$cultivar, cex=0.7)

plot(log_norm_ct_none_rnipals@scores[,2], log_norm_ct_none_rnipals@scores[,3], col=samplelist$cultivar, pch=19)
text(log_norm_ct_none_rnipals@scores[,2], log_norm_ct_none_rnipals@scores[,3], labels=samplelist$cultivar, cex=0.7)

plot(log_norm_ct_none_rnipals@scores[,3], log_norm_ct_none_rnipals@scores[,4], col=samplelist$cultivar, pch=19)
text(log_norm_ct_none_rnipals@scores[,3], log_norm_ct_none_rnipals@scores[,4], labels=samplelist$cultivar, cex=0.7)


# pareto scaling, rnipals
# trial
palette(brewer.pal(6, "Paired"))
func_5pairs_plot(log_norm_ct_pareto_rnipals, samplelist, "trial", 19, "trial effect (pareto scaling, rnipals)")
func_pca_plot(log_norm_ct_pareto_rnipals, 1, 2, samplelist, "trial", 19, 1, 
              "bottomright", 0.8, maintext="trial effect (pareto scaling, rnipals)")

# cultivar
palette(rainbow(34))
func_5pairs_plot(log_norm_ct_pareto_rnipals, samplelist, "cultivar", 19, "cultivar effect (pareto scaling, rnipals)")
func_pca_plot(log_norm_ct_pareto_rnipals, 1, 4, samplelist, "cultivar", 19, 1, 
              "bottomright", 0.8, maintext="cultivar effect (pareto scaling, rnipals)")

plot(log_norm_ct_pareto_rnipals@scores[,1], 
     log_norm_ct_pareto_rnipals@scores[,5], 
     col=samplelist$cultivar, pch=19, xlab="PC1", ylab="PC5")
text(log_norm_ct_pareto_rnipals@scores[,1], 
     log_norm_ct_pareto_rnipals@scores[,5], 
     labels=samplelist$cultivar, cex=0.7)


# no scaling, ppca
# trial
palette(brewer.pal(6, "Paired"))
func_5pairs_plot(log_norm_ct_none_ppca, samplelist, "trial", 19, "trial effect (no scaling, ppca)")

func_pca_plot(log_norm_ct_none_ppca, 1, 2, samplelist, "trial", 19, 1, 
              "bottomright", 0.8, maintext="trial effect (no scaling, ppca)")

# cultivar
palette(rainbow(34))
func_5pairs_plot(log_norm_ct_none_ppca, samplelist, 
                 "cultivar", 19, "cultivar effect (no scaling, ppca)")
# separation of some cultivars in PC4 vs. PC5
func_pca_plot(log_norm_ct_none_ppca, 4, 5, samplelist, "cultivar", 19, 1, 
              "bottomleft", 0.8, maintext="cultivar effect (no scaling, ppca)")

plot(log_norm_ct_none_ppca@scores[,4], 
     log_norm_ct_none_ppca@scores[,5], 
     col=samplelist$cultivar, pch=19)

text(log_norm_ct_none_ppca@scores[,4], 
     log_norm_ct_none_ppca@scores[,5], 
     labels=samplelist$cultivar, cex=0.7)

plot(log_norm_ct_none_rnipals@scores[,1], 
     log_norm_ct_none_rnipals@scores[,2], 
     col="grey", pch=19)

text(log_norm_ct_none_rnipals@scores[,1], 
     log_norm_ct_none_rnipals@scores[,2], 
     labels=samplelist$sampleID, cex=0.7)
dev.off()

palette("default")
```


# Correlation plot
```{r corrplot}
cor_log_norm_ct <- cor(log_norm_ct,use="complete")
write.table(cor_log_norm_ct, "output/breeder/cor_log_norm_ct.txt", sep="\t")

pdf("figures/breeder/cor_log_norm_ct.pdf", width=10, height=10)
corrplot(cor_log_norm_ct, order="hclust", tl.col="black", tl.cex=0.6)
dev.off()
```


# Remove Factors by ANOVA models using R script by Jan Lisec
```{r normalize (Remove Factors)}
levels(samplelist$trial)
levels(samplelist$treatment) # NO TREATMENT
levels(samplelist$cultivar)

log_norm_ct_2 <- apply(log_norm_ct, 2, RemoveFactors, sam=samplelist, 
                              facs=c("cultivar", "trial"), 
                              keep=c("cultivar"))
```


## PCA after Remove Factors
```{r PCA after Remove Factors}

log_norm_ct_pareto_rnipals_2 <- func_prep_pca(log_norm_ct_2, scale_method = "pareto", center_option = FALSE, 
                                             pc_number = 5, pca_method = "rnipals")

pdf("figures/breeder/pca_log_norm_ct_2.pdf")

# year effect
palette(heike_palette_2)
pairs(log_norm_ct_pareto_rnipals_2@scores[,1:5], 
      col = samplelist$year, pch = 19)

plot(log_norm_ct_pareto_rnipals_2@scores[,1], 
     log_norm_ct_pareto_rnipals_2@scores[,2], 
     col = samplelist$year, pch = 19, 
     xlab = "PC1 (27.5%)", ylab = "PC2 (7%)")
legend("bottomleft", levels(samplelist$year), fill=1:6)


# trial effect
palette(brewer.pal(6, "Paired"))
pairs(log_norm_ct_pareto_rnipals_2@scores[,1:5], 
      col = samplelist$trial, pch = 19)

plot(log_norm_ct_pareto_rnipals_2@scores[,1], 
     log_norm_ct_pareto_rnipals_2@scores[,2], 
     col = samplelist$trial, pch = 19, 
     xlab = "PC1 (27.5%)", ylab = "PC2 (7%)")
legend("bottomleft", levels(samplelist$trial), fill=1:6)


# cultivar effect
palette(rainbow(34))
pairs(log_norm_ct_pareto_rnipals_2@scores[,1:5], 
      col=samplelist$cultivar, pch=19)

plot(log_norm_ct_pareto_rnipals_2@scores[,3], 
     log_norm_ct_pareto_rnipals_2@scores[,4], 
     col = samplelist$cultivar, pch=19, 
     xlab = "PC3 (6.8%)", ylab = "PC4 (5.8%)")

text(log_norm_ct_pareto_rnipals_2@scores[,3], 
     log_norm_ct_pareto_rnipals_2@scores[,4], 
     labels = samplelist$cultivar, cex=0.7)

plot(log_norm_ct_pareto_rnipals_2@scores[,4], 
     log_norm_ct_pareto_rnipals_2@scores[,5], 
     col = samplelist$cultivar, pch=19, 
     xlab = "PC4 (5.8%)", ylab = "PC5 (5%)")

text(log_norm_ct_pareto_rnipals_2@scores[,4], 
     log_norm_ct_pareto_rnipals_2@scores[,5], 
     labels = samplelist$cultivar, cex=0.7)

dev.off()
palette("default")
```


# ANOVA with 2 factors
```{r ANOVA with 2 factors}
res_anova_adj <- func_anova_2fac(normalized_values = log_norm_ct, 
                                 trial_factors = samplelist, 
                                 factor1 = "trial", 
                                 factor2 = "cultivar",
                                 threshold = 0.01,
                                 analyte_names = colnames(log_norm_ct))

head(res_anova_adj)
```


# Boxplots per gene
```{r boxplots per gene}
pdf("figures/breeder/boxplots_per_gene_per_cultivar.pdf", 8, 6)
func_boxplot_1fac(normalized_values = log_norm_ct, 
                  trial_factors = samplelist, 
                  factor = "cultivar", 
                  res_anova_adj = res_anova_adj, 
                  cols = cols_cultivar_34, 
                  analyte_names = colnames(log_norm_ct))
dev.off()

pdf("figures/breeder/boxplots_per_gene_per_trial.pdf", 8, 6)
par(mar=c(7, 4.1, 7, 2.1))
func_boxplot_1fac(normalized_values = log_norm_ct, 
                  trial_factors = samplelist, 
                  factor = "trial", 
                  res_anova_adj = res_anova_adj, 
                  cols = brewer.pal(6, "Paired"), 
                  analyte_names = colnames(log_norm_ct))
dev.off()
```


# Save workspace and sessioninfo
```{r save workspace}
save.image("qpcr_data_breeder.RData")
sessionInfo()
```


